backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job_schedule/service/commands/PickingJobScheduleAutoAssignCommand.java

```diff
@@ -0,0 +1,343 @@
+import de.metas.shipping.CarrierProductId;
+import de.metas.util.Check;
+import org.adempiere.service.ISysConfigBL;
+import org.adempiere.warehouse.WarehouseId;
+import de.metas.handlingunits.shipmentschedule.api.IHUShipmentScheduleBL;
+import de.metas.workplace.Workplace;
+import de.metas.workplace.WorkplaceId;
+import de.metas.workplace.WorkplaceRepository;
+import lombok.Builder;
+import lombok.NonNull;
+
+import javax.annotation.Nullable;
+import java.math.BigDecimal;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static de.metas.handlingunits.picking.job_schedule.service.commands.CreateOrUpdatePickingJobSchedulesCommand.SYSCONFIG_CARRIER_PRODUCT_REQUIRED;
+
+public class PickingJobScheduleAutoAssignCommand
+{
+	// Services
+	@NonNull private final WorkplaceRepository workplaceRepository;
+	@NonNull private final ShipmentScheduleRepository shipmentScheduleRepository;
+	@NonNull private final PickingJobScheduleRepository pickingJobScheduleRepository;
+	@NonNull private final IHUShipmentScheduleBL shipmentScheduleBL;
+	@NonNull private final ISysConfigBL sysConfigBL;
+	@NonNull private final ProductRepository productRepository;
+
+	// Params
+	@NonNull private final PickingJobScheduleAutoAssignRequest request;
+
+	@Builder
+	private PickingJobScheduleAutoAssignCommand(
+			@NonNull final WorkplaceRepository workplaceRepository,
+			@NonNull final ShipmentScheduleRepository shipmentScheduleRepository,
+			@NonNull final PickingJobScheduleRepository pickingJobScheduleRepository,
+			@NonNull final IHUShipmentScheduleBL shipmentScheduleBL,
+			@NonNull final ISysConfigBL sysConfigBL,
+			@NonNull final ProductRepository productRepository,
+			@NonNull final PickingJobScheduleAutoAssignRequest request)
+	{
+		this.workplaceRepository = workplaceRepository;
+		this.shipmentScheduleRepository = shipmentScheduleRepository;
+		this.pickingJobScheduleRepository = pickingJobScheduleRepository;
+		this.shipmentScheduleBL = shipmentScheduleBL;
+		this.sysConfigBL = sysConfigBL;
+		this.productRepository = productRepository;
+		this.request = request;
+	}
+
+	public void execute()
+	{
+		final ImmutableList<Workplace> workplacesWithCapacity = workplaceRepository.getAllActive()
+				.stream()
+				.filter(workplace -> workplace.getMaxPickingJobs() > 0)
+				.collect(ImmutableList.toImmutableList());
+
+		if (workplacesWithCapacity.isEmpty())
+		{
+			return;
+		}
+
+		final ImmutableSet<WarehouseId> warehouseIds = workplacesWithCapacity.stream()
+				.map(Workplace::getWarehouseId)
+				.collect(ImmutableSet.toImmutableSet());
+
+		final ImmutableList<ShipmentSchedule> allEligibleShipmentSchedules = shipmentScheduleRepository.getBy(
+				ShipmentScheduleQuery.builder()
+						.warehouseIds(warehouseIds)
+						.preparationDate(request.getPreparationDate())
+						.fromCompleteOrderOrNullOrder(true)
+						.includeWithQtyToDeliverZero(false)
+						.includeProcessed(false)
+						.build())
+				.stream()
```

Comment:
how much data could be here?

Reply:
Still have this concern

> Concern if adding limit:
we hit a spot where we split 2 shipment schedules of the same order in 2 batches, making totalQtyToDeliverByOrderId incorrect

Reply:
documented